docs/proposals/config_data.md

```diff
@@ -0,0 +1,262 @@
+```json
+{
+    "apiVersion": "v1",
+    "name": "PodConfiguration",
+    "kind": "ConfigData",
+    "data": {
+        "API_SERVER": "10.240.13.14:8998",
+        "API_VERSION": "v1",
+        "DB_HOST": "10.240.55.3",
+        "DB_PORT": "3306"
+    }
+}
+```
+
+To persist this ConfigData object to the API server, use the following command:
+```bash
+kubectl create -f pod-config.json
+```
+
+Now, the ``PodConfiguration`` ConfigData object is stored in ``etcd``.
+The new object can be accessed, referenced, used and updated through the API server.
+
+## Configuration Workflow
+
+The component configuration workflow is given below:
+
+1. ``ConfigData`` object is created through the API server.
+2. New components using the ``ConfigData`` object are created.
+3. ``ConfigData`` object is retrieved using the downward API.
+4. ``Data`` key/value pairs of ``ConfigData`` will be exposed as either environment variables (``EnvVarSource``) or as a new volume (``ConfigDataVolumeSource``).
+5. API server is watched for modifications to the object.
+6. After a modification, actions 3, 4, and 5 are repeated.
+
+**Note**: Environment variables specified in ``ConfigData`` will override the environment variables
+with the same name found in the system if consumed as environment variables.
+
+## Use Cases
+
+#### Pod Configuration
+
+```json
+{
+    "apiVersion": "v1",
+    "name": "PodConfiguration1",
+    "kind": "ConfigData",
+    "data": {
+        "API_SERVER": "10.240.13.14:8998",
+        "API_VERSION": "v1",
+        "DB_HOST": "10.240.55.3",
+        "DB_PORT": "3306"
+    }
+}
+```
+``PodConfiguration1`` is intended to be consumed as environment variable key/value pairs.
+
+```json
+{
+    "apiVersion": "v1",
+    "name": "PodConfiguration2",
+    "kind": "ConfigData",
+    "data": {
+        "config.yaml": "CONFIG_YAML_STRING_CONTENTS",
+        "spec.yaml": "SPEC_YAML_STRING_CONTENTS"
+    }
+}
+```
+``PodConfiguration2`` is intended to be used as a mounted volume containing two files with their respective contents.
+
+Then, configuration objects are persisted to the API server as follows:
+```bash
+kubectl create -f pod-configuration-1.json 
+kubectl create -f pod-configuration-2.json
+```
+
+We create a replication controller specification that creates pods using these two configuration objects:
+```yaml
+apiVersion: v1
+kind: ReplicationController
+metadata:
+  name: ExampleReplicationController
+spec:
+  replicas: 2
+  selector:
+    app: dummy-app
+  template:
+    metadata:
+      labels:
+        app: dummy-app
+    spec:
+      # New pods will be created using the ConfigData objects.
+      # Each config object will be updated be it a volume or environment variables.
+      # But, some of them might require a restart on update.
+      # Note: If multiple env ConfigData objects are specified, let's call them Env1 and Env2.
+      #       If both of them contains a same named environment variable, Env2 will be the final one.
```

Comment:
This proposal is certainly flexible enough to allow that, for sure. There's no reason you couldn't have multiple ConfigData objects being used for various env vars via the downward API. But we still want to be explicit with the mapping.